### PR TITLE
GSoC 2017 - FEniCS blog post

### PR DESCRIPTION
**Pull request by ysimillides**
Blog Post for GSOC 2017 (FEniCS.jl)

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Let's get this registered first.

File: blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+---
+
+<script type="text/javascript"
+  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
+</script>
+
+
+# Introduction
+
+Throughout this Google Summer of Code project I, along with my mentors, aimed to create a Wrapper for the FEniCS Finite Element Toolbox in the Julia Language. Our work done can be found at [FEniCS.jl](https://github.com/JuliaDiffEq/FEniCS.jl) . This would allow users to perform FEM calculations directly in Julia, utilizing our PyCall.jl wrapping functionality. We currently have wrapped the main  functionality, along with providing the necessary instructions to add further components when they are deemed necessary. Members of the Julia community not directly related to the project also contributed small fixes and suggestions throughout the project. The majority of the code produced has already been merged to the GitHub repository (which was created specifically for this project). One of the main improvements which would greatly increase its usage would be the further integration with the JuliaDiffEq package.
```
Comment (ChrisRackauckas) [inline blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: Link PyCall.jl

Comment (ysimillides) [reply blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: Done

File: blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+# What is FEniCS?
+
+**[FEniCS](https://fenicsproject.org/)** describes itself as a popular open-source (LGPLv3) computing platform for solving partial differential equations (PDEs). FEniCS enables users to quickly translate scientific models into efficient finite element code. In layman's terms it lets us solve math problems.
+
+# But, how does it work? 
+
+The finite element method is a numerical method that solves partial differential equations by discretizing the domain (breaking it up into small pieces, generally triangles called finite elements) and then solving the resulting equations (which are generally a system of algebraic equations) using calculus of variations to approximate our required solution. This is nearly impossible to do by hand, so computer software is required to aid in solving it. Our wrapper provides calls to the meshing functionality, the assembly of the stiffness matrices and the solution of the variational problems. It also provides access to various *helper* functions, that make usage easier.
```
Comment (ChrisRackauckas) [inline blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: Might want to explain this more thoroughly. Use some pictures? It solves the weak form by doing a Galerkin approximation of the function into some sparse basis. This is usually something like splitting up the domain into triangles, and representing the function via the values at the nodes of the triangles. You can show a picture for this meshing. Then assembly is re-interpreting the operator equations in terms of this basis of triangles.

File: blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+# Challenges 
+
+Due to the nature of this project, which relied on FEniCS, we faced various challenges throughout the summer. These included, but where not limited to build errors, where various parts of the package failed to compile, to unexpected errors in the usability of the code. Chris and Bart where very helpful, in both pointing these out, and in assisting in fixing them. In some parts the documentation was slightly patchy which also complicated parts of the project as some functions where ambiguous towards their intended use. 
+
+# Further Improvements
+
+I hope to be able to maintain and improve the package, using it where possibly throughout my further studies. Some identifiable improvements, in order of difficulty are : 
+
+- Fixing precompilation which would provide a large performance benefit. This error is well documented, and the fix is relatively simple. At the same time it would require the rewriting of a large segment of the codebase due to the way we currently access functions and attributes.
```
Comment (ChrisRackauckas) [inline blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: Link to the issues.

**Comment by ChrisRackauckas** [comment]
@barche any comments?

File: blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+# What is FEniCS?
+
+**[FEniCS](https://fenicsproject.org/)** describes itself as a popular open-source (LGPLv3) computing platform for solving partial differential equations (PDEs). FEniCS enables users to quickly translate scientific models into efficient finite element code. In layman's terms it lets us solve math problems.
+
```
Comment (barche) [inline blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: The last sentence here is a bit too general, maybe something like "it lets us write complicated mathematical equations and solves them automatically using computer simulation"

File: blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+vtkfile = File('poisson/solution.pvd')
+vtkfile << U.pyobject #exports the solution to a vtkfile
+
+```
+Apart from just defining the problem, we can also access and save the arrays corresponding to various variational forms. We can do this as follows :
+
+```julia
+a = dot(grad(u),grad(v))*dx #this sets up the variatonal form from the previous problem
+variational_matrix = get_array(a)
+
```
Comment (barche) [inline blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: Maybe specify here what the compatibility of the returned type is with Julia, e.g. is it an `AbstractArray`?

File: blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+I hope to be able to maintain and improve the package, using it where possibly throughout my further studies. Some identifiable improvements, in order of difficulty are : 
+
+- Fixing precompilation which would provide a large performance benefit. This error is well documented, and the fix is relatively simple. At the same time it would require the rewriting of a large segment of the codebase due to the way we currently access functions and attributes. 
+- Improving plotting. We currently rely on the FEniCS plotting backend to plot the necessary functions/meshes/objects. For more detailed visualization we can use Paraview, like in FEniCS. A direct Julia plotter would be nice, as we could provide further customization to our objects.
+- Integration with JuliaDiffEq. We can currently specify and create the necessary objects for the solution of some FEM problems. We have also provided interfaces for accessing most of their attributes aswell as exporting the necessary arrays. Despite this, we currently have no automatic way of seamlessly accessing them via other packages. By providing this access, we would be able to greatly extend the packages capabilities.
+
+
```
Comment (barche) [inline blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: Maybe add a comment here about the FEniCS form compiler, which takes e.g. a matrix assembly expression and compiles this to C code and then machine code. Clearly, it would be an opportunity to replace this whole chain directly with Julia.

**Comment by barche** [comment]
@ChrisRackauckas Thanks for the heads-up, almost missed this.
@ysimillides Good job!

File: blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md (hunk 2 above)
Comment (ysimillides) [reply blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: implemented most of the changes, remains to add various mesh plots

File: blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md (hunk 5 above)
Comment (ChrisRackauckas) [reply blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: @ysimillides address this and I'll approve

**Comment by ViralBShah** [comment]
Good to go?

File: blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md (hunk 5 above)
Comment (ysimillides) [reply blog/_posts/GSoC 2017 project A Wrapper for the FEniCS Finite Element Toolbox.md]: has been added.
